File: aws/VerifyAuthChallenge.py
import boto3
import json
import jwt
import re
import base64
import urllib.request


from os import environ


LOGINID_BASE_URL = environ.get("LOGINID_BASE_URL") or ""
LOGINID_SECRET_NAME = environ.get("LOGINID_SECRET_NAME") or ""

secretsmanager = boto3.client("secretsmanager")


def lambda_handler(event: dict, _: dict) -> dict:
    print(event)
    request, response = event["request"], event["response"]
    metadata = request["clientMetadata"]
    authentication_type = metadata.get("authentication_type")
    challenge_answer = request["challengeAnswer"]

    # if authentication type is empty throw error
    if not authentication_type:
        raise Exception("Authentication type not found")

    # pass for next round to obtain authentication type
    if challenge_answer == "AUTH_PARAMS":
        response["answerCorrect"] = False
        return event

    try:
        valid_types = ["FIDO2_CREATE", "FIDO2_GET", "JWT_ACCESS"]

        if valid_types.count(authentication_type):
            # parse JWT token
            username = event["userName"]
            payload = jwt.decode(
                challenge_answer,
                options={"verify_signature": False},
            )

            if payload.get("aud") != get_loginid_app_id():
                raise Exception("Invalid JWT token")
            if payload.get("username") != username:
                raise Exception("Invalid JWT token")

            verify_jwt_access_token(challenge_answer)

        elif authentication_type == "EMAIL_OTP":
            # get otp from private challenge parameters
            otp = request["privateChallengeParameters"].get("otp", "")
            if otp != challenge_answer:
                raise Exception("Invalid OTP")

        else:
            raise Exception("Invalid authentication type")

        response["answerCorrect"] = True

        return event
    except Exception as e:
        print(e)
        response["answerCorrect"] = False
        return event


def get_key_id() -> str:
    secret = secretsmanager.get_secret_value(SecretId=LOGINID_SECRET_NAME)
    key_id = secret["SecretString"]
    return key_id


def get_loginid_app_id() -> str:
    pattern = r"https://([0-9a-fA-F-]+)\.api\..*\.loginid\.io"
    match = re.search(pattern, LOGINID_BASE_URL)

    if match:
        app_id = match.group(1)
    else:
        raise Exception("Invalid LoginID base URL")

    return app_id


def verify_jwt_access_token(token: str) -> None:
    print(token)
    url = f"{LOGINID_BASE_URL}/fido2/v2/mgmt/token/verify"
    headers = {
        "Content-Type": "application/json",
    }
    payload = {
        "jwtAccess": token
    }

    # basic auth
    credentials = f"{get_key_id()}:"
    encoded_credentials = base64.urlsafe_b64encode(credentials.encode()).decode()
    authorization = f"Basic {encoded_credentials}"
    headers["Authorization"] = authorization

    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers=headers)

    with urllib.request.urlopen(req) as res:
        if res.status != 204:
            raise Exception("Invalid JWT token")

    return

File: aws/nodejs/VerifyAuthChallenge.mjs
/**
 * Copyright 2024 LoginID
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import https from "https";
import { Buffer } from "buffer";
import {
  GetSecretValueCommand,
  SecretsManagerClient,
} from "@aws-sdk/client-secrets-manager";

const LOGINID_BASE_URL = process.env.LOGINID_BASE_URL || "";
const LOGINID_SECRET_NAME = process.env.LOGINID_SECRET_NAME || "";

const secretsmanager = new SecretsManagerClient();

export const handler = async (event) => {
  console.log(event);
  const { request, response } = event;
  const { clientMetadata, challengeAnswer } = request;
  const { authentication_type: authenticationType } = clientMetadata;

  // if authentication type is empty throw error
  if (!authenticationType) {
    throw new Error("Authentication type not found");
  }

  // pass for next round to obtain authentication type
  if (challengeAnswer === "AUTH_PARAMS") {
    response.answerCorrect = false;
    return event;
  }

  try {
    switch (authenticationType) {
      case "FIDO2_CREATE":
      case "FIDO2_GET":
      case "JWT_ACCESS":
        // parse JWT token
        const payload = parseJwt(challengeAnswer);

        if (payload.aud !== getLoginIdAppId()) {
          throw new Error("Invalid JWT token");
        }
        if (payload.username !== event.userName) {
          throw new Error("Invalid JWT token");
        }

        await verifyJwtAccessToken(challengeAnswer);
        break;

      case "EMAIL_OTP":
        // get otp from private challenge parameters
        const otp = request?.privateChallengeParameters?.otp;
        if (otp !== challengeAnswer) {
          throw new Error("Invalid OTP");
        }
        break;

      default:
        throw new Error("Invalid authentication type");
    }

    response.answerCorrect = true;
    return event;
  } catch (e) {
    console.log(e);
    response.answerCorrect = false;
    return event;
  }
};

const getKeyId = async () => {
  const secret = await secretsmanager.send(
    new GetSecretValueCommand({ SecretId: LOGINID_SECRET_NAME })
  );
  return secret.SecretString;
};

const getLoginIdAppId = () => {
  const pattern = /https:\/\/([0-9a-fA-F-]+)\.api\.(.*\.)?loginid\.io/;
  const match = LOGINID_BASE_URL.match(pattern);

  if (match) {
    return match[1];
  } else {
    throw new Error("Invalid LoginID base URL");
  }
};

const verifyJwtAccessToken = async (token) => {
  const url = `${LOGINID_BASE_URL}/fido2/v2/mgmt/token/verify`;
  const payload = JSON.stringify({ jwtAccess: token });

  const keyId = await getKeyId();
  const credentials = `${keyId}:`;
  const encodedCredentials = Buffer.from(credentials).toString("base64");
  const options = {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Basic ${encodedCredentials}`,
    },
  };

  return new Promise((resolve, reject) => {
    const req = https.request(url, options, (res) => {
      let data = "";

      res.on("data", (chunk) => {
        data += chunk;
      });

      res.on("end", () => {
        if (res.statusCode === 204) {
          resolve();
        } else {
          console.log("Error response:", data);
          reject(new Error("Invalid JWT token"));
        }
      });
    });

    req.on("error", (e) => {
      reject(e);
    });

    req.write(payload);
    req.end();
  });
};

const parseJwt = (token) => {
  const [_, payload] = token.split(".");
  const decodedPayload = Buffer.from(payload, "base64").toString("utf-8");
  return JSON.parse(decodedPayload);
};
